src/environments.js

```diff
@@ -0,0 +1,104 @@
+var ParseError = require("./ParseError");
+
+/*
+ * An environment is modelled by a constructor which has an instance
+ * of EnvironmentBase as its prototype.  It may implement a method
+ * begin(), the return value of which is ignored, and must implement a
+ * method end(), the return value of which is just like the return
+ * value of a function, i.e. should have a "type" property.
+ *
+ * These two methods may make use of various properties which will be
+ * set on the object by the parser:
+ * - body          the parsed body, as an array
+ * - args          an array of arguments to the \begin call
+ * - argPositions  the positions associated with the \begin and its arguments
+ * - parser        the corresponding Parser instance
+ * - lexer         the corresponding Lexer instance
+ * - beginNode     the ParseResult describing the \begin function result
+ * - endNode       the ParseResult describing the \end function result
+ * Most of these are mostly useful for error reporting.
+ *
+ * An environment may specify many of the same properties that a
+ * function may, like e.g. numArgs.  These will be applied to the
+ * \begin invocation, to capture environment arguments.
+ */
+
+function EnvironmentBase() {}
+EnvironmentBase.prototype = {
+    numArgs: 0,
+    greedingess: 1,
+    allowedInText: false,
+    numOptionalArgs: 0,
+    begin: function() { /* default is no operation, end does the work. */ }
+};
+
+function array() { }
+array.prototype = new EnvironmentBase();
+array.prototype.numArgs = 1;
+array.prototype.end = function() {
```

Comment:
I read through this a little to quickly on the first take.  I see that you're creating new EnvironmentBase() objects so this should be okay.